include/caffe/datum_DB_factory.hpp

```diff
@@ -0,0 +1,98 @@
+#ifndef CAFFE_DATUMDB_FACTORY_H_
+#define CAFFE_DATUMDB_FACTORY_H_
+
+#include <map>
+#include <string>
+
+#include "caffe/common.hpp"
+#include "caffe/proto/caffe.pb.h"
+
+namespace caffe {
+
+class DatumDB;
+
+class DatumDBRegistry {
+ public:
+  typedef DatumDB* (*Creator)(const DatumDBParameter&);
+  typedef std::map<string, Creator> CreatorRegistry;
+  typedef std::map<string, DatumDB* > SourceRegistry;
+
+  static CreatorRegistry& Registry() {
+    static CreatorRegistry* g_registry_ = new CreatorRegistry();
+    return *g_registry_;
+  }
+
+  static SourceRegistry& Sources() {
```

Comment:
I think this is here to prevent multiple opens of the same DB? If so, it seems like an odd place for that... what I expect in a `*Registry` class is code dealing with the static-variable registration trick, but database paths come at run time. I think if we need this it would make more sense in the DB class itself (or its subclasses, if we need it for only some types of DBs).